adapter/internal/ingress/client_manager.go

```diff
@@ -64,21 +65,29 @@ func NewClientManager(connector ConnectionBuilder, connCount int, ttl time.Durat
 func (c *ClientManager) Next() v2.EgressClient {
 	for {
 		idx := atomic.AddUint64(&c.nextIdx, 1)
-		conn := (*connection)(atomic.LoadPointer(&c.connections[idx%uint64(len(c.connections))]))
-		if conn.client != nil {
+
+		c.mu.RLock()
+		actualIdx := int(idx % uint64(len(c.connections)))
```

Comment:
I was looking at the other place where we do `len(c.connections)` in `monitorConnectionsForRolling` and was wondering if they need to be protected since we are not modifying the length of the slice. I ran a test and the only time `len` needs to be protected is if you are changing the slice, for instance:

```go
package main

func main() {
	a := []int{1, 2, 3}
	go func() {
		println(len(a))
	}()
	a = append(a, 4)
}
```

results in:

```
==================
WARNING: DATA RACE
Read at 0x00c420076000 by goroutine 5:
  main.main.func1()
      /Users/jasonkeene/gobal/src/lenrace/main.go:6 +0x3b

Previous write at 0x00c420076000 by main goroutine:
  main.main()
      /Users/jasonkeene/gobal/src/lenrace/main.go:8 +0x16a

Goroutine 5 (running) created at:
  main.main()
      /Users/jasonkeene/gobal/src/lenrace/main.go:7 +0xec
==================
4
Found 1 data race(s)
```

Since we allocate a slice in the constructor with a fixed size that never changes these reads of `len(c.connections)` are safe to do without synchronization.

Incidentally, the same thing applies to maps. You can safely read length if you do not modify it. Though, in most situations this is not the case and even if you are only writing into a key that existed in the map before concurrent access you still get a data race.